[all] Improve msg_api and logging of message

damienmarchal opened the pull request:
This is the return of PR https://github.com/sofa-framework/sofa/pull/106 taking into account all the feedback and discussion in https://github.com/sofa-framework/sofa/issues/179

This PR contains: 
 -  a PerComponentLoggingMessageHandler to stores the messages in components in a much cleaner way compared to the existing code in Sofa.
 -  the old history code is removed but a compatibility layer on top of the new one is provided to 
      provide backward compatibility (eg in third party code like qtSofaQuick).  
 -  the runSofa GUI is slighly updated and improved to make use of the new component. 
 -  the msg_api is extended to facilitate its use and coherency
 -  I fixed all around in the code base the mis-usages of printLog eg: if(f_printLog) cout << ....  
 -  I didn't replaced the f_printLog data field because it would have broken the API too much 
      (maybe one day if someone care)

Now a bried summary of the new functions in the msg_*  API
```cpp
   msg_info() << "This is a message sended by the this component" ;
   msg_info(other) << "This is a message sended by the 'other' component" ;
   msg_info("A String") << "This message is not sended by a component";
```
Each version also exist in a conditional form eg:
```cpp 
    msg_info_when(!file.valid()) << "File is not valid" 
```

In the previous examples you can also replace 'info' with the type of message you want to emit among:
'info', 'advice', 'deprecated', 'warning', 'error', 'fatal'. 
  
'warning', 'error' and 'fatal' and 'deprecated' message cannot be muted with the printLog parameter while 'info','advice',' can. 

______________________________________________________
<!--- Please leave this at the end of your message -->
This PR: 
- [x] builds with SUCCESS for all platforms on the CI.
- [x] does not generate new warnings nor unit test failures.
- [x] does not break existing scenes.
- [x] does not break API compatibility.
- [x] has been reviewed.
- [x] is more than 1 week old.

**Reviewers will merge only if all these checks are true.**

damienmarchal committed:
[SofaKernel] ADD a function and storage for the messages.

The serr/sout are storing the message as an unique string,
each message being concatenated to the previous one.
This is:
   - inefficient
   - loose information (number of message as well as the messages informations)
   - in case of too much message the behavior is to prevent adding more message instead of dropping the
     old ones.


This commit add the code needed to really store the message.

damienmarchal committed:
[SofaKernel] In ComponentInfo now store the a Base* instead of the component name.

damienmarchal committed:
[SofaKernel] MOVE RichConsoleStyleMessageFormatter from helper to core.

Some handler and formatter rely on the Base* object. This means that
they cannot be in helper unless we link core to helper.

The general idea is probably to move the complete msg_api from helper
to core (but this is something left for future).

damienmarchal committed:
[SofaKernel] FIX compilation problem because of wrong location for RichStyle.

damienmarchal committed:
[SofaKernel] ADD a new message handler to store the logging history per sofa component.

When this handler is installed the messages passing through are
logged in a circular buffer into the sofa components (Base).
Not distinction is made on the type of message to log.

If you want to implement distinct message to log please consider chaining this handler
with the RoutingMessageHandler.

damienmarchal committed:
[SofaKernel] ADD to Base the needed methods & attribute to log the messages.

damienmarchal committed:
[SofaKernel]: Fix PerComponentLoggingMessageHandler.

damienmarchal committed:
[SofaKernel] ADD new test for the logging of message in components.

damienmarchal committed:
[SofaKernel] ADD data fields for fine grain control of the message log.

'messageLevel' should be use to specify above which level a message is emitted by the component.
'loggingLevel' should be use to specify above which level a message is loggued in the component.
'loggingSize' should be use to specify the size of the internal logging queue.

I'm not fully convinced of that.

damienmarchal committed:
[SofaKernel] FIX a minor queue size bug.

damienmarchal committed:
[SofaKernel] ADD a new test on the logging queue.

damienmarchal committed:
[SofaKernel] FIX logging_test and queue size behavior.

damienmarchal committed:
[SofaKernel] REMOVE the ugly #ifdef WIN32

damienmarchal committed:
[SofaKernel] ADD SOFA_CORE_API to avoid link error on windows.

damienmarchal committed:
Merge remote-tracking branch 'upstream/master' into improve-messageapi

damienmarchal committed:
[CHANGELOG] Update the changelog.

damienmarchal committed:
Merge remote-tracking branch 'upstream/master' into improve-messageapi

damienmarchal committed:
[SofaKernel] Remove the un-needed Data<> in Base.h

damienmarchal committed:
[SofaKernel] Forgot to finish the removal the un-needed Data<> in base.

damienmarchal committed:
[SofaKernel] Update the test to fix a minor regression due to the removing of the Data field.

damienmarchal committed:
[All] Remove the old output & warning buffers all around.

the old string based buffer are now replaced with a proper API
to log & access message in component.

This new API is now used in the runSofa GUI to improve
the messages in the panel.

The Output & Warning panels have been removed into a single one
because it is much better (I can explain but I'm too tired).

I remove the clearWarning and clearOutput from some visitors...
Again I could explain but I'm tired.

damienmarchal committed:
[runSofa] Small UX improvement related to messages.

damienmarchal committed:
[SofaKernel] Restore Base.*

damienmarchal committed:
[SofaKernel] Add implicit version of the messaging API

New functions:
  msg_info() << "Eg"
  msg_warning() << "Eg"
  msg_error() << "Eg"
  msg_fatal() << "Eg"
  msg_advice() << "Eg"
  msg_deprecated() << "Eg"

  msg_info_when(a==b) << "Eg"
  msg_warning_when(a==b) << "Eg"
  msg_error_when(a==b) << "Eg"
  msg_fatal_when(a==b) << "Eg"
  msg_advice_when(a==b) << "Eg"
  msg_deprecated_when(a==b) << "Eg"

  msg_info_when(a==b, "toto") << "Eg"
  msg_warning_when(a==b, "toto") << "Eg"
  msg_error_when(a==b, "toto") << "Eg"
  msg_fatal_when(a==b, "toto") << "Eg"
  msg_advice_when(a==b, "toto") << "Eg"
  msg_deprecated_when(a==b, "toto") << "Eg"

  msg_info_when(a==b, this) << "Eg"
  msg_warning_when(a==b, this) << "Eg"
  msg_error_when(a==b, this) << "Eg"
  msg_fatal_when(a==b, this) << "Eg"
  msg_advice_when(a==b, this) << "Eg"
  msg_deprecated_when(a==b, this) << "Eg"


As well as the same with dmsg_*

damienmarchal committed:
[SofaKernel] Add InfoMessageLevel in Base & GUI.

damienmarchal committed:
[SofaKernel] remove the use of f_printLog in Base

damienmarchal committed:
[SofaKernel] WIP on msg_*

damienmarchal committed:
[SofaKernel] Remove a trivial warning.

damienmarchal committed:
[runSofa] Add a way for users to suggests messages improvements via a web page.

damienmarchal committed:
[all] replace the mis-usage of f_printLog with msg_info

damienmarchal committed:
[SofaKernel] Remove the d_msgInfoLevel attribute

It is too early to add this attribute.

damienmarchal committed:
[SofaKernel] Remove the use of printLog

damienmarchal committed:
[all] replace a lot of if(f_printLog) with msg_* version

damienmarchal committed:
Merge remote-tracking branch 'upstream/master' into improve-messageapi

damienmarchal committed:
[SofaKernel] restore the initial existing.txt file i accidentally commited.

damienmarchal committed:
[SofaKernel] Still fixing the existing.txt file

damienmarchal committed:
Update existing.txt

damienmarchal committed:
[runSofa] fix the compilation problem with the old version of Qt

damienmarchal committed:
Merge branch 'improve-messageapi' of github.com:SofaDefrost/sofa into improve-messageapi

damienmarchal committed:
[runSofa] fix compilation problem related with old version of Qt

damienmarchal committed:
[SofaKernel] Fix some problem with the dmsg_* that was printing too much messages.


Also remove an experimental feature asking for user to suggest
better error messages.

thomas-lemaire reviewed (commented):
Hi @damienmarchal 
You are so brave (english meaning) !
I see you have replaced the "sendl" by "  \n"
Couldn't it break windows message output ?
Also, but it is more cosmetics, this more common in c++ to use a (std::)endl for this purpose

damienmarchal commented:
Hi @thomas-lemaire,

About the "\n" ... why would it breaks windows console printing ? Windows and linux does not have the same "\n" but a windows app emits the "\n" the windows system understand.

About using "\n" or better std::endl. My target was to supports messages written with a simple markdown syntax for which the "\n" is not an explicit line break, it is "  \n" (double-space before the \n) . So i replaced std::endl/sendl by this "  \n" 
But I found that ugly and hard to read so I finally made 'msgendl' that contains the explicit markdown line break, but I forgot to remove the old manually placed "  \n" I used all around. 

Now I will have to re-fix the old "  \n" all around (but this is nothing :)).

Do you think msgendl is ok ? 
Do you like the msgendl ?

The change on how the messages are printed in runSofa is visible here: 
![newmessage](https://cloud.githubusercontent.com/assets/12644882/23545144/72df7122-fffa-11e6-8c53-fc9a4415c493.png)

The two panels Output/Error are now merged into a single "Message" one and the panel now shows in its title the number of messages (in this PR I removed the column with "suggest a better message)

In the Graph view there is also a small alert (green, yellow, red) on top of the icons showing the kind of message an object have. 

EDIT: I fixed the "  \n"

damienmarchal committed:
[all] Replace the "  \n" with msgendl

damienmarchal committed:
[all] Fix a minor compilation problem in SVDLinearSOlver

damienmarchal committed:
[sofaKernel] Replace \n with msgendl

thomas-lemaire commented:
ok, I agree with your choice ! 
thomas 

----- Mail original -----

> De: "Damien Marchal" <notifications@github.com>
> À: "sofa-framework/sofa" <sofa@noreply.github.com>
> Cc: "thomas-lemaire" <thomas.lemaire@inria.fr>, "Mention"
> <mention@noreply.github.com>
> Envoyé: Vendredi 3 Mars 2017 10:18:55
> Objet: Re: [sofa-framework/sofa] [all] Improve msg_api and logging of message
> (#190)

> Hi @thomas-lemaire ,

> About the "\n" ... why would it breaks windows console printing ? Windows and
> linux does not have the same "\n" but a windows app emits the "\n" the
> windows system understand.

> About using "\n" or better std::endl. My target was to supports messages
> written with a simple markdown syntax for which the "\n" is not an explicit
> line break, it is " \n" . So i replaced std::endl/sendl by " \n"
> But I found that ugly and hard to read so I finally made 'msgendl' that
> contains the explicit markdown line break, but I forgot to remove the old
> manually placed " \n" I used all around.

> Now I will have to re-fix the old " \n" all around (but this is nothing :)).

> Do you think msgendl is ok ?
> Do you like the msg_endl

> The change on how the messages are printed in runSofa is visible here:

> The two panels are merged and the panel now shows in its title the number of
> messages.
> (in this PR I removed the column with "suggest a better message)

> In the Graph view there is now a small alert (green, yellow, red) on top of
> th icons showing the kind of message an object have.

> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub , or mute the thread .

damienmarchal committed:
[all] Replace std::cerr with calls to msg_* api

damienmarchal committed:
Merge branch 'improve-messageapi' of github.com:SofaDefrost/sofa into improve-messageapi

damienmarchal committed:
[SofaKernel] Fix compilation on windows.

bcarrez committed:
Update tests to comply with new "printLog" policy

damienmarchal committed:
Merge branch 'master' into improve-messageapi

damienmarchal committed:
[SofaKernel] Fix compilation merge

damienmarchal committed:
Merge branch 'master' into improve-messageapi

matthieu-nesme commented on SofaKernel/framework/framework_test/resources/existing.txt:
```diff
@@ -198,4 +198,4 @@
 }#include<TODOD> int main(int argc...){ ... }
-}
\ No newline at end of file
+}
```
il traine encore ce fichier ? :/

damienmarchal replied:
OUps :)

matthieu-nesme committed:
Update RichConsoleStyleMessageFormatter.cpp

be careful with string copies.
(this is the first time I use the online editor, it looks goood)

matthieu-nesme commented on SofaKernel/framework/sofa/core/objectmodel/Base.h:
```diff
@@ -332,17 +334,13 @@ class SOFA_CORE_API Base
-
     /// effective ostringstream for logging
     mutable std::ostringstream _serr, _sout;
+    mutable std::deque<sofa::helper::logging::Message> m_messageslog ;
```
So the component is still storing its own messages?
Is it only for this dev step?

damienmarchal replied:
Actually I was thinking you asked to implement this feature otherwise the msg_* api wouln't be equivalent enough to the sout/serr one. 

I personally have no clear idea whether or not we could in some future remove the per component log. 

So this PR is more about improving the implementation of the existing approach and improving the way sofa handle message. 

Gven the way msg_* is designed removing the per object backlog and replace it with a single history wouldn't be a hard if someone is willing to do that.

matthieu-nesme commented on SofaKernel/framework/sofa/core/objectmodel/Base.h:
```diff
@@ -524,14 +527,22 @@ class SOFA_CORE_API Base
 
 } // namespace sofa
 
+/// This allow an object to be muted.
```
is it really necessary?

damienmarchal replied:
I think so, the notMuted is used internally by msg_info().

matthieu-nesme commented on SofaKernel/framework/sofa/helper/logging/Message.h:
```diff
@@ -113,6 +126,8 @@ class SOFA_HELPER_API Message
 
     /// possible levels of messages (ordered)
     enum Type {Info=0, Advice, Deprecated, Warning, Error, Fatal, TEmpty, TypeCount};
+    typedef std::set<Type> TypeSet;
```
what is the idea of that?

damienmarchal replied:
To write things like that: 
```
    std::string infos = base->getLoggedMessagesAsString( { Message::Info,
                                                           Message::Advice,
                                                           Message::Deprecated } );
```

matthieu-nesme replied:
Ok I see, is there a way to fill "Message::AnyTypes" automatically from "enum Type", it seems really easy to have incoherences.

damienmarchal replied:
Isn't what line 42 in Message.cpp is doing ? 
```cpp 
Message::TypeSet Message::AnyTypes = {Type::Info,Type::Advice,Type::Deprecated,
                                      Type::Warning,Type::Error,Type::Fatal};
```

matthieu-nesme commented on applications/plugins/Compliant/numericalsolver/EigenSparseResponse.inl:
```diff
@@ -40,15 +40,14 @@ void EigenSparseResponse<LinearSolver,symmetric>::factor(const rmat& H) {
 
-    if( this->f_printLog.getValue() ) serr<<"factor: "<<H<<sendl;
-
+    msg_info() << "factor: ";
```
is it really the same? H is missing and I have to look "muted" more in detail.

damienmarchal replied:
H is missing...thanks for pointing. 
And it cannot be the same because in the discussion related to this PR we decided that warning and error cannot be muted with the printLog. 
So it can either be: 
    msg_info() << "factor " ;  (in this case it preserve the mutability via the printLog) 
or  
    msg_warning() << "factor"; (in this case  it preserve the "Warning" level of the message)

As the message was not really informative on why it was a warning (no text describing what what wrong, how to remove the warning etc...).  I though it was more a kind of general information...and thus convert it to msg_info.

matthieu-nesme commented on applications/plugins/Compliant/numericalsolver/SequentialSolver.cpp:
```diff
@@ -305,46 +305,41 @@ void BaseSequentialSolver::solve_impl(vec& res,
-        serr << "iterations: " << k << ", (abs) residual: " << (net - mapping_response * lambda).norm() << sendl;
-	
-
+    msg_info() << "iterations: " << k << ", (abs) residual: " << (net - mapping_response * lambda).norm() << sendl;
```
missing muted again?

damienmarchal replied:
No...msg_info contains the test for mutability.

matthieu-nesme commented on applications/plugins/Compliant/odesolver/CompliantImplicitSolver.cpp:
```diff
@@ -549,11 +549,9 @@ using namespace core::behavior;
-            sout << "constraint size n: " <<sys.n<< sendl;
-        }
+
+        dmsg_info() << "dynamics size m: " << sys.m << msgendl
```
muted?

damienmarchal replied:
dmsg_info contains the test for printLog.

damienmarchal committed:
[Compliant] Fix missing "H" in a print.

damienmarchal committed:
Merge branch 'improve-messageapi' of github.com:SofaDefrost/sofa into improve-messageapi

damienmarchal committed:
[all] Remove the call to std::cerr all around.

matthieu-nesme commented:
I like the PR, but I am concern about the modifs in Compliant/Flexible.
I totally agree we did not use the printLog flag the right way, but the PR cannot be merged as it as it will create troubles in those plugins.
What would be the right way to keep the same semantic we wanted?

damienmarchal commented:
Hi @matthieu-nesme 

Thanks for your feedback, I really did my best on it to avoid as much as possible trouble.

I suggest to cut this PR in several part (haha) to remove the Flexible/Compliant files from the changes, this woud make the move much smooth. 

This will give you time to point out specifically the most problematic points in which you think that the remplacement will not do the work so that I can propose solutions to be closer to the "semantic you want". 

I looked into the Flexible/Compliant and it seems to me that only a dozen of replacement (out of hundreds) are not equivalent to the old code. 

If you take time pointing me which one are causing you troubles I will gladely help in how fixing them to use the msg_api AND and preserving the existing behavior.  To convince you that this is worth the effort I took examples from Flexible/Compliant to illustrate the benefit of the changes.  
====  ====
```cpp
- if(this->f_printLog.getValue()==true)
-          std::cout << SOFA_CLASS_METHOD << std::endl;
+         dmsg_info() << SOFA_CLASS_METHOD ;
```
- Is functionnaly equivalent to the old code (if the application have installed an handler routing the message to std::cout)
- It will print message in a totally consistant way with other sofa componant (which very important for users)
- It is shorter to write (very important for the developer :)), 
- It benefit from the message dispatcher/handler (so it can be logger or routed or sended or whatever GUI in the end user application)
- All developpers immediately know that the message is not for the end user of Sofa but for a developers. Consequently the message can be very drafty. It can also be remove by setting SOFA_DEVTOOLS to false and thus it should have a zero cost in a release  

A second example: 
```cpp
- if(this->f_printLog.getValue()) if(pos.size())    
-        std::cout<<"TopologyGaussPointSampler: "<< pos.size() <<" generated samples"<<std::endl;
+msg_info_when(pos.size()!=0) << " topologyGaussPointSampler: "<< pos.size() <<" generated samples" ;
```
- Is functionnaly equivalent to the old code
- Shorter to write
- Consistant behavior with other component,
- It benefit from the message dispatcher/handler (so it can be logger or routed or sended into Backlog depending on the end user application)

And one with an error:
```cpp
     if( inv.info() != Eigen::Success ){
-        std::cerr << SOFA_CLASS_METHOD<<"non invertible block Schur." << std::endl;
-        std::cerr << schur << std::endl;
+        msg_error(SOFA_CLASS_METHOD) << "non invertible block Schur." << msgendl
+                                     << schur ;
```
- Shorter to write  (using msg_error_when() it would have to be even shorter)
- Consistant way to present message  to the user
- All the text is considered is a single message with a single call to msg_* 
- Is functionnaly equivalent to the old code 
- Uou can write tests that check that the message
  is emitted (ExpectMessage/MessageAsTestFailure) so it become possible to test how components behave when they are "mis-used".... while with the old code it was just 'not done' at all.

damienmarchal committed:
[SofaKernel] Add mutex to the entry point of the msg_* API

damienmarchal committed:
[Compliant] Fix a missing .str() in a message

damienmarchal committed:
[Flexible&Compliant] Restore the old way to handle messages.

damienmarchal commented:
Et voilà, no more change in Flexible & Compliant.

For discussion the last revision continaing them is this one.
https://github.com/sofa-framework/sofa/pull/190/commits/50d6ccd23f6f52e22518710f6ef8187294a562bd

damienmarchal commented:
[ci-build]

damienmarchal committed:
Merge remote-tracking branch 'upstream/master' into improve-messageapi

damienmarchal committed:
[SofaKernel] Fix one test in Logging_text.

The counter was not rested before the test so it was failing
taking into accout messages emitted in other tests.

damienmarchal commented:
[ci-build]

damienmarchal committed:
Merge remote-tracking branch 'upstream/master' into improve-messageapi

# Conflicts:
#	CHANGELOG.md

damienmarchal committed:
Merge branch 'improve-messageapi' of github.com:SofaDefrost/sofa into improve-messageapi

damienmarchal commented:
[ci-build]

damienmarchal committed:
Fix the static mutex problem that crash on windows.

damienmarchal committed:
Merge branch 'master' into improve-messageapi

damienmarchal committed:
[SofaTest] Restore the old text in Regression_test I screwed during the merge

damienmarchal committed:
[SofaKernel] WIP Fix the VS2013 problem with static mutex

VS2013 is not conformant with the c++11 standard regarding static
thread initialization. This cause a lot of mess in my code that I'm
trying to fix.

damienmarchal committed:
[SofaKernel] Makes the mutex non static because of VS2013

damienmarchal committed:
[SofaKernel] Restore the initial behavior.

damienmarchal committed:
[SofaKernel] Refactor to move the 'Base' related function from sofa::helper::logging to sofa::core::objectmodel.

damienmarchal committed:
Merge remote-tracking branch 'upstream/master' into improve-messageapi

# Conflicts:
#	modules/SofaVolumetricData/DistanceGrid.cpp

damienmarchal committed:
[SofaKernel] FIX link problem on VS.

damienmarchal committed:
[SofaKernel] Doc in msg_api

damienmarchal committed:
[SofaKernel] Add minimalistic comments in the classes.

damienmarchal committed:
[SofaKernel] ADD missing file lost during previous cleaning.

damienmarchal committed:
[SofaKernel] Add the OpenIssue suffix to the Logging_test about threading.

This test will succeed when the Logging implementation will be thread safe.
It is easy to do in c++x11 but because of VS2013 it is not that easy.
So I will make the implementation thread safe when we will drop VS2013
support.

damienmarchal commented:
[ci-build]
